### PR TITLE
FALCON-910 

### PR DESCRIPTION
**Pull request by cryptoe**
Not able to upload patch to review board.
Raising pull request for review. FALCON-910

### CODE REVIEW

File: common/src/main/java/org/apache/falcon/entity/ClusterHelper.java [hunk 1]
```diff
@@ -117,10 +118,10 @@ private static String getNormalizedUrl(Cluster cluster, Interfacetype type) {
         return normalizedPath.substring(0, normalizedPath.length() - 1);
     }
 
-    public static String getLocation(Cluster cluster, String locationKey) {
+    public static Location getLocation(Cluster cluster, ClusterLocationType clusterLocationType) {
         for (Location loc : cluster.getLocations().getLocations()) {
-            if (loc.getName().equals(locationKey)) {
-                return loc.getPath();
+            if (loc.getName().value().equals(clusterLocationType.value())) {
```
Comment (shwethags) [inline common/src/main/java/org/apache/falcon/entity/ClusterHelper.java]: Use enum equals - loc.getName() == clusterLocationType

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 2]
```diff
@@ -243,37 +244,72 @@ private void validateLocations(Cluster cluster) throws ValidationException {
...
+                throw new ValidationException(
+                        "Unable to validate the location of name: " + stagingLocation.getName().value() + " with path: "
+                                + stagingLocation.getPath() + " for cluster " + cluster.getName(), e);
+            }
+            if (ClusterHelper.getLocation(cluster, ClusterLocationType.WORKING) == null) {
+                //Creating location type of working in the sub dir of staging dir with perms 755. FALCON-910
+                Location workingLocation = new Location();
+                workingLocation.setName(ClusterLocationType.WORKING);
+                String workingDirPath = stagingLocation.getPath();
+                workingDirPath = (workingDirPath.endsWith("/")
```
Comment (shwethags) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: User fs Path()

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 3]
```diff
@@ -243,37 +244,72 @@ private void validateLocations(Cluster cluster) throws ValidationException {
...
-            if (locationName.equals("temp")) {
-                continue;
-            }
+        if (ClusterHelper.getLocation(cluster, ClusterLocationType.STAGING) == null) {
+            throw new ValidationException(
+                    "Unable to find the mandatory location of name: " + ClusterLocationType.STAGING.value()
+                            + " for cluster " + cluster.getName());
+        } else {
 
+            Location stagingLocation = ClusterHelper.getLocation(cluster, ClusterLocationType.STAGING);
```
Comment (shwethags) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: You can call ClusterHelper.getLocation(cluster, ClusterLocationType.STAGING) once and re-use later

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 4]
```diff
@@ -243,37 +244,72 @@ private void validateLocations(Cluster cluster) throws ValidationException {
...
             try {
-                checkPathOwnerAndPermission(cluster.getName(), location.getPath(), fs,
-                        "staging".equals(locationName)
-                                ? HadoopClientFactory.ALL_PERMISSION
-                                : HadoopClientFactory.READ_EXECUTE_PERMISSION);
+                checkPathOwnerAndPermission(cluster.getName(), stagingLocation.getPath(), fs,
+                        HadoopClientFactory.READ_EXECUTE_PERMISSION, false);
```
Comment (shwethags) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: Shouldn't the staging be ALL permissions

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 5]
```diff
@@ -243,37 +244,72 @@ private void validateLocations(Cluster cluster) throws ValidationException {
...
+            } else {
+                Location workingLocation = ClusterHelper.getLocation(cluster, ClusterLocationType.WORKING);
+                if (stagingLocation.getPath().equals(workingLocation.getPath())) {
+                    throw new ValidationException(
+                            "Location with name: " + stagingLocation.getName().value() + " and " + workingLocation
+                                    .getName().value() + " cannot of same path: " + stagingLocation.getPath()
+                                    + " for cluster :" + cluster.getName());
+                } else {
+                    try {
+                        checkPathOwnerAndPermission(cluster.getName(), workingLocation.getPath(), fs,
```
Comment (shwethags) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: checkPathOwnerAndPermission() should catch IOException and throw FalconException. IOException is a transient failure and can't be ValidationException

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 6]
```diff
@@ -281,17 +317,28 @@ private void checkPathOwnerAndPermission(String clusterName, String location, Fi
...
+            if (fileStatus.getPermission().toShort() != expectedPermission.toShort()) {
+                LOG.error("Location {} has permissions {}, should be {}", locationPath, fileStatus.getPermission(),
+                        expectedPermission);
+                throw new ValidationException(
+                        "Path " + locationPath + " has permissions: " + fileStatus.getPermission() + ", should be "
+                                + expectedPermission);
+            }
+        } else {
+            if (fileStatus.getPermission().toShort() < expectedPermission.toShort()) {
+                LOG.error("Location {} has permissions {}, should be {}", locationPath, fileStatus.getPermission(),
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: If you are logging and throwing an exception with very similar message. Create the message once and re-use it save a  redundant lines of code.

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java (hunk 4 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: https://issues.apache.org/jira/browse/FALCON-910?focusedCommentId=14227619&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14227619

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java (hunk 3 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: acked

File: common/src/main/java/org/apache/falcon/entity/ClusterHelper.java (hunk 1 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/ClusterHelper.java]: acked

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java (hunk 5 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: acked

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java (hunk 2 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: acked

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java (hunk 6 above)
Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: acked

File: common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java [hunk 7]
```diff
@@ -243,58 +244,95 @@ private void validateLocations(Cluster cluster) throws ValidationException {
...
-            try {
-                checkPathOwnerAndPermission(cluster.getName(), location.getPath(), fs,
-                        "staging".equals(locationName)
-                                ? HadoopClientFactory.ALL_PERMISSION
-                                : HadoopClientFactory.READ_EXECUTE_PERMISSION);
-            } catch (IOException e) {
-                throw new ValidationException("Unable to validate the location of name: "
-                        + location.getName() + " with path:" + location.getPath()
-                        + " for cluster " + cluster.getName(), e);
+        if (stagingLocation == null) {
```
Comment (sriksun) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: From the older version of the code, it seems like staging perms need to be ALL and data perms need to be Read/Execute. But this seems to be changing the permission scheme. Can you confirm ?

Comment (cryptoe) [inline common/src/main/java/org/apache/falcon/entity/parser/ClusterEntityParser.java]: Yup I have changed the perms check from 777 (All perms) to atleast 755 for staging perms as per
https://issues.apache.org/jira/browse/FALCON-910?focusedCommentId=14227619&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14227619 

Regarding perms this what i am doing:
https://issues.apache.org/jira/browse/FALCON-910?focusedCommentId=14255793&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14255793

**Comment by cryptoe** [comment]
closing pull request as its merged

**Pull request closed without merging**